### DISCUSSION TITLE
Renovate for Docker does not suggest update after a certain image version through Artifactory

### DISCUSSION THREAD

**Discussion by renan-alm** [renovatebot/renovate#20275] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.125.1

### If you're self-hosting Renovate, select which platform you are using.

github.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

GitHub (.com and Enterprise Server

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Wanted end result.

Renovate suggesting any version of Grafana over `docker/grafana/grafana:9.2.6`

### What you tried so far.

- Using several configurations.
- Upgrading to version `9.3.0` to see if it would suggest me a later existing one, for example, `9.3.1`(which again, exists).
- Downgrading below `9.2.6` and running Renovate towards Artifactory, which gives me a positive result but not the latest. 
- Downgrading below `9.2.6`, changing the image to go directly to Dockerhub, and running Renovate, which gives me the latest update. 

### Relevant debug logs

<details><summary>Positive upgrade</summary>

Note the field `updates`.

````
DEBUG: packageFiles with updates (repository=XZY/renovate-renan, baseBranch=master)
       "config": {
         "dockerfile": [
           {
             "packageFile": "Dockerfile",
             "deps": [
               {
                 "depName": "XZY.jfrog.io/docker/grafana/grafana",
                 "currentValue": "9.1.0",
                 "replaceString": "XZY.jfrog.io/docker/grafana/grafana:9.1.0",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "final",
                 "depIndex": 0,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "9.2.6",
                     "newValue": "9.2.6",
                     "newMajor": 9,
                     "newMinor": 2,
                     "updateType": "minor",
                     "branchName": "renovate/docker-deps"
                   }
                 ],
                 "warnings": [],
                 "versioning": "docker",
                 "registryUrl": "https://XYZ.jfrog.io",
                 "currentVersion": "9.1.0",
                 "isSingleVersion": true,
                 "fixedVersion": "9.1.0"
               }
             ]
           }
         ],
````
</details>

<details><summary>Bad case - Not finding latest versions</summary>

Note the field `updates`.

````
DEBUG: packageFiles with updates (repository=XYZ/renovate-renan, baseBranch=master)
       "config": {
         "dockerfile": [
           {
             "packageFile": "Dockerfile",
             "deps": [
               {
                 "depName": "XYZ.jfrog.io/docker/grafana/grafana",
                 "currentValue": "9.2.7",
                 "replaceString": "XYZ.jfrog.io/docker/grafana/grafana:9.2.7",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "final",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "versioning": "docker",
                 "registryUrl": "https://XXYZ.jfrog.io",
                 "currentVersion": "9.2.7",
                 "fixedVersion": "9.2.7"
               }
             ]
           }
         ],
````

</details>

<details><summary>Good case - Not through Artifactory</summary>

````
DEBUG: packageFiles with updates (repository=XYZ/renovate-renan, baseBranch=master)
       "config": {
         "dockerfile": [
           {
             "packageFile": "Dockerfile",
             "deps": [
               {
                 "depName": "grafana/grafana",
                 "currentValue": "9.2.7",
                 "replaceString": "grafana/grafana:9.2.7",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "final",
                 "depIndex": 0,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "9.3.6",
                     "newValue": "9.3.6",
                     "newMajor": 9,
                     "newMinor": 3,
                     "updateType": "minor",
                     "branchName": "renovate/docker-deps"
                   }
                 ],
                 "warnings": [],
                 "versioning": "docker",
                 "registryUrl": "https://index.docker.io",
                 "currentVersion": "9.2.7",
                 "isSingleVersion": true,
                 "fixedVersion": "9.2.7"
               }
             ]
           }
         ],
````

</details>

**Comment by szpak** [comment]
I had similar case with Harbor used as a proxy to DockerHub. Renovate made a call:
> https://my-harbor/v2/proxy/library/maven/tags/list?n=10000

(should be visible in the debug logs)

and Harbor was returning the locally available tags. No extra hit to DockerHub to fetch all the tags available upstream. When I explicitly requested a given version/tag in some other build, it was fetched and the another Renovate call would return it as well. Maybe Artifactory does the same?

In fact, it seems to be on the registry/proxy side. I plan to ask the Harbor guys if tag list call could/should be forwarded upstream (and eventually cached for some time). However, it also brings some issues (such as a greater chance to hit the rate limits).

**Reply by szpak** [reply]
FYI. I asked about that on the Harbor forum, and it seems to be already improved in Harbor 2.5 - https://github.com/goharbor/harbor/discussions/18187.

**Reply by viceice** [reply]
getting the tags list from docker hub currently doesn't count to any limits. only getting any JSON manifest or binary blob does count.

**Reply by viceice** [reply]
@HonkingGoose can you add a note to the docker datasource docs about harbor proxy cache  and version 2.5.0+ requirements?

**Reply by HonkingGoose** [reply]
Here you go! 😉 

- #20322

**Comment by viceice** [comment, marked as answer]
harbor and artifactory are known to not properly passing the upstream tags list which renovate needs to discover available versions. there's nothing renovate can do here. this was already discussed multiple times here.

---

harbor proxy cache feature works since v2.5.0 with renovate. so make sure to upgrade.
